### DISCUSSION TITLE
Division by zero at time 0 despite all values > 0

### DISCUSSION THREAD

**Discussion by bflo96** [OpenModelica/OpenModelica#12456] [Q&A]
Hello everyone,

In my first model, I have this:

```
Real Req[nC,nR];
Real auxR[nC,nE];
(...)
for c in 1:nC loop
  for r in 1:nR loop
    Req[c,r] = 1 / sum(auxR[c,e] for e in pos[r,1]:pos[r,2]);
  end for;
end for;
```
pos[r,1] and pos[r,2] are previously defined and correspond to the "limits" of each Req[c,r].
For example, I have nE = 104, but I only want to use elements 1 to 5 to calculate Req[c,1]. The number of elements I want to use for each Req[c,r] is different.

This model works just fine when I run it on its own.

However, I want to use these results in a second model. It might look repetitive, but this makes it easier for me to later construct the problem I want. Then, in the second model I extend the previous and:
```
for d in 1:nD loop
    for j in 1:nR loop
      D[d].R_b[j].R = Req[d, j];
      D[d].R_t[j].R = Req[d+1, j];
    end for;
end for;
```

But I get the following error, which I don't understand why it happens. Req is calculated correctly through the first model.

> division by zero at time 0, (a=1) / (b=0), where divisor expression is sum (auxR[1,e] for e in posElem[1,1]:posElem[1,2]

I feel like I am missing something obvious here.

Thank you.

**Comment by casella** [comment]
@bflo96 it would be nice to have an MWE that we can run ourserlves.

**Reply by bflo96** [reply]
Sorry for the late reply. Thank you so much for your response!

I managed to find a way to solve my problem, even though I haven't figured out why it happens.
From the tests I perform, I got the feeling it was some sort of initialization problem, so I managed to solve it by defining

`Req[c,r] = 1 / sum(if auxR[c,e] == 0 then 0.0001 else auxR[c,e] for e in pos[r,1]:pos[r,2]);`

It has no impact on my results and it seems it helps the calculations to move forward... 

**EDIT:** Actually, a simpler solution was the give a start value to the variable auxR (for example start=1). 
Is it a common and good practice to give a start value to all variables?

**Reply by casella** [reply]
Good question. You need to give a reasonable start value to the variables that end up in a nonlinear implicit system during initialization. In fact, only the iteration variables, if tearing is used. Unfortunately, in many cases it's hard to tell which these variables will be, because it depends on how you connect components with each other. After working on this for a while, you develop a 6th sense on it.

In any case, OMC gives you a warning about variables showing up in implicit equations and not having explicitly set start attributes. I guess you saw them but did not know what to do with them. One option is to check the solution sequence with the declarative debugger.

This is in fact one dark side of equation-based, object-oriented modelling. There are ways to mitigate it: better tool diagnostic messages, the declarative debugger, homotopy-based initialization, etc.

**Reply by casella** [reply]
On what constitutes a reasonable initial guess for the Newton solver, you can check our paper https://doi.org/10.1016/j.amc.2021.125991. There is also an open version of it on ArXiv https://arxiv.org/abs/1911.12433

**Reply by casella** [reply]
On homotopy-based initialization, please check https://ep.liu.se/ecp/063/010/ecp11063010.pdf and http://www.ep.liu.se/ecp/063/011/ecp11063011.pdf

**Reply by casella** [reply]
On the declarative (a.k.a. equation-based) debugger, please check https://ep.liu.se/ecp/096/020/ecp14096020.pdf and https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/debugger.html#the-equation-based-debugger

**Reply by bflo96** [reply]
Thank you! I'll take a look into what you shared.
